File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ElmConditionRequirement.java
package org.cqframework.cql.elm.requirements;

import org.hl7.elm.r1.Expression;
import org.hl7.elm.r1.VersionedIdentifier;

public class ElmConditionRequirement extends ElmExpressionRequirement {
    public ElmConditionRequirement(VersionedIdentifier libraryIdentifier, Expression expression, ElmPropertyRequirement property, ElmExpressionRequirement comparand) {
        super(libraryIdentifier, expression);

        if (property == null) {
            throw new IllegalArgumentException("property is required");
        }
        this.property = property;

        if (comparand == null) {
            throw new IllegalArgumentException("comparand is required");
        }
        this.comparand = comparand;
    }

    protected ElmPropertyRequirement property;
    public ElmPropertyRequirement getProperty() {
        return this.property;
    }

    protected ElmExpressionRequirement comparand;
    public ElmExpressionRequirement getComparand() {
        return this.comparand;
    }
}

File: Src/java/elm-fhir/src/main/java/org/cqframework/cql/elm/requirements/ElmConditionRequirement.kt
package org.cqframework.cql.elm.requirements

import org.hl7.elm.r1.Expression
import org.hl7.elm.r1.VersionedIdentifier

open class ElmConditionRequirement(
    libraryIdentifier: VersionedIdentifier,
    expression: Expression,
    property: ElmPropertyRequirement,
    comparand: ElmExpressionRequirement,
) : ElmExpressionRequirement(libraryIdentifier, expression) {
    var property: ElmPropertyRequirement?
        protected set

    var comparand: ElmExpressionRequirement?
        protected set

    init {
        this.property = property

        this.comparand = comparand
    }

    val isTargetable: Boolean
        get() =
            comparand != null &&
                (comparand!!.isLiteral ||
                    comparand!!.isTerminologyReference ||
                    comparand!!.isParameterReference)
}
